src/Components/Server/src/Circuits/DefaultCircuitFactory.cs

```diff
@@ -38,16 +40,17 @@ internal class DefaultCircuitFactory : CircuitFactory
             _logger = _loggerFactory.CreateLogger<DefaultCircuitFactory>();
         }
 
-        public override CircuitHost CreateCircuitHost(
-            HttpContext httpContext,
+        public CircuitHost CreateCircuitHost(
+            string serializedComponentRecords,
             CircuitClientProxy client,
             string baseUri,
             string uri,
             ClaimsPrincipal user)
         {
-            // We do as much intialization as possible eagerly in this method, which makes the error handling
-            // story much simpler. If we throw from here, it's handled inside the initial hub method.
-            var components = ResolveComponentMetadata(httpContext);
+            if (!DescriptorSerializer.TryDeserializeComponentDescriptorCollection(serializedComponentRecords, out var components))
```

Comment:
Does it make sense to do this in the caller? IIRC this method has one caller. We added a new dependency here just so we could throw an exception from this method.

This seems like it would gel really well with the [other validations](https://github.com/aspnet/AspNetCore/blob/master/src/Components/Server/src/ComponentHub.cs#L90).

I also like the current invariant that this method doesn't fail, and doesn't have input that needs validation.

Reply:
Ok, I've moved it out.